Model Element sometimes jump when applying JS Transforms after StageMode

Phipson opened the pull request:
#### 6025199d730f1bd4c4a6afa86822a329d5e32b09
<pre>
Model Element sometimes jump when applying JS Transforms after StageMode
<a href="https://bugs.webkit.org/show_bug.cgi?id=289993">https://bugs.webkit.org/show_bug.cgi?id=289993</a>
<a href="https://rdar.apple.com/145795564">rdar://145795564</a>

Reviewed by Ada Chan.

This PR addresses an issue where the entityTransform after stagemode is desynced from the actual transform
of the model, causing it to &quot;jump&quot; when applying JS transforms thereafter. The problem is two-fold:
1. The driverInitialized boolean, which determines whether we forward updates to the entityTransform, was incorrectly set.
2. The animation transform subscriptions were not correctly reseting the driver&apos;s internal state machine, causing subsequent
stagemode interactions to not correctly report the entityTransform back to the model element.

The fix involves changing the definition of driverInitialized such that it reports transform changes during and after
the gesture is released, up until the animation(s) are complete. It also adds a check such that we only report the entityTransform
via the animation only when the gesture ends. This prevents the internal state machine from entering a bad state if one animation
finishes before the other or get cancelled because the user re-initiates the gesture mid-animation.

* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
(WKStageModeInteractionDriver.allowAnimationObservation):
(WKStageModeInteractionDriver.pitchAnimationCompletionSubscription):
(WKStageModeInteractionDriver.yawAnimationCompletionSubscription):
(WKStageModeInteractionDriver.stageModeInteractionInProgress):
(WKStageModeInteractionDriver.interactionDidBegin(_:)):
(WKStageModeInteractionDriver.interactionDidEnd):
(WKStageModeInteractionDriver.operationDidUpdate(_:)):
(WKStageModeInteractionDriver.subscribeToPitchChanges):
(WKStageModeInteractionDriver.subscribeToYawChanges):

Canonical link: <a href="https://commits.webkit.org/292429@main">https://commits.webkit.org/292429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/575235a31eee8e2b93b55d195a4ad117d1c73f96

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95935 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5495 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100996 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46444 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15844 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23981 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73148 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30372 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98938 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11863 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53473 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11597 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45779 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81763 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103023 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23002 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82187 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23253 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82677 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81549 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20472 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26138 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/16339 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22965 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28120 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22624 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26104 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24365 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -180,72 +175,98 @@ final public class WKStageModeInteractionDriver: NSObject {
         // Settle the pitch of the interaction container
         pitchSettleAnimationController = self.interactionContainer.move(to: initialTargetPose, relativeTo: self.interactionContainer.parent, duration: kPitchSettleAnimationDuration, timingFunction: .easeOut)
         subscribeToPitchChanges()
+        pitchAnimationCompletionSubscription = self.interactionContainer.scene?.subscribe(to: AnimationEvents.PlaybackCompleted.self, on: self.interactionContainer) { [weak self] event in
```
It doesn't look like `event` is referenced. Should it be just `_`?

achan00 commented on Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
```diff
@@ -180,72 +175,98 @@ final public class WKStageModeInteractionDriver: NSObject {
         // based on the user's current orbit velocity
         yawDecelerationAnimationController = self.turntableAnimationProxyEntity.move(to: Transform(scale: .one, rotation: .init(ix: 0, iy: 0, iz: 0, r: 1), translation: .init(repeating: 1)), relativeTo: nil, duration: kMaxDecelerationDuration, timingFunction: .linear)
         subscribeToYawChanges()
+        yawAnimationCompletionSubscription = self.turntableAnimationProxyEntity.scene?.subscribe(to: AnimationEvents.PlaybackTerminated.self, on: self.turntableAnimationProxyEntity) { [weak self] event in
```
Ditto.

achan00 replied:
Curious why the yaw one subscribes to `PlaybackTerminated` while the pitch one subscribes to `PlaybackCompleted`?

Phipson replied:
So for the deceleration animation, we are not directly applying the animation onto the entity that's decelerating. Instead, we only listen to the animation updates to dampen the entity based on the drag velocity. If the velocity is so small it reaches a certain threshold, we stop the animation immediately. This means that the animation can end before it completes. For the pitch animation, we always wait for the animation to complete. This is why we have `PlaybackTerminated` for the yaw animation and `PlaybackCompleted` for pitch animation.

Phipson committed:
Model Element sometimes jump when applying JS Transforms after StageMode
https://bugs.webkit.org/show_bug.cgi?id=289993
rdar://145795564

Reviewed by Ada Chan.

This PR addresses an issue where the entityTransform after stagemode is desynced from the actual transform
of the model, causing it to "jump" when applying JS transforms thereafter. The problem is two-fold:
1. The driverInitialized boolean, which determines whether we forward updates to the entityTransform, was incorrectly set.
2. The animation transform subscriptions were not correctly reseting the driver's internal state machine, causing subsequent
stagemode interactions to not correctly report the entityTransform back to the model element.

The fix involves changing the definition of driverInitialized such that it reports transform changes during and after
the gesture is released, up until the animation(s) are complete. It also adds a check such that we only report the entityTransform
via the animation only when the gesture ends. This prevents the internal state machine from entering a bad state if one animation
finishes before the other or get cancelled because the user re-initiates the gesture mid-animation.

* Source/WebKit/WebKitSwift/StageMode/WKStageMode.swift:
(WKStageModeInteractionDriver.allowAnimationObservation):
(WKStageModeInteractionDriver.pitchAnimationCompletionSubscription):
(WKStageModeInteractionDriver.yawAnimationCompletionSubscription):
(WKStageModeInteractionDriver.stageModeInteractionInProgress):
(WKStageModeInteractionDriver.interactionDidBegin(_:)):
(WKStageModeInteractionDriver.interactionDidEnd):
(WKStageModeInteractionDriver.operationDidUpdate(_:)):
(WKStageModeInteractionDriver.subscribeToPitchChanges):
(WKStageModeInteractionDriver.subscribeToYawChanges):

Canonical link: https://commits.webkit.org/292429@main
